### ISSUE TITLE
Legend on the heatmap

### ISSUE THREAD

**Issue by a-velt** [Alanocallaghan/d3heatmap#6]
Hi,

I tried to use your package in order to add a legend on my heatmap. Here my code : 

`d3heatmap(data_to_plot_heatmap_MANUAL_samples_log,show_color_legend=TRUE, colors = colorRampPalette(rev(brewer.pal(n = 7, name = "RdYlBu")))(100),cexCol=1,cexRow=1, labCol=colnames(data_to_plot_heatmap_MANUAL_samples_log))`

Here my output : 
![image](https://user-images.githubusercontent.com/23554476/35979520-d8c24e28-0ce8-11e8-880c-e9ab03797159.png)

I see a "legend" but, I have no idea of the numerical value assigned at each color. In your example you have : 
![image](https://user-images.githubusercontent.com/23554476/35979583-f97b9f3e-0ce8-11e8-9b62-12c539d4d081.png)

With values assigned. 

Do you know what's wrong with my code?

Thank you.
Best,
Amandine

**Comment by Alanocallaghan** [comment]
Hi Amandine, 

I will have a look at this issue this evening. 

In the meantime, you might be interested in the package `heatmaply`, which I also contribute to. It accomplishes many of the same tasks as d3heatmap, and is available [on CRAN](https://cran.r-project.org/web/packages/heatmaply/index.html) and [on github](https://github.com/talgalili/heatmaply)

Thanks for the interest
Alan

**Comment by a-velt** [comment]
Ok thank you, I know heatmaply but how can I put the output it in my shiny app ? Because it creates a shiny app separately, in my memories.

**Comment by talgalili** [comment]
You could look at the shinyHeatmaply package an example.

On Feb 8, 2018 5:51 PM, "Amandine Velt" <notifications@github.com> wrote:

Ok thank you, I know heatmaply but how can I put the output it in my shiny
app ? Because it creates a shiny app separately, in my memories.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/Alanocallaghan/d3heatmap/issues/6#issuecomment-364154432>,
or mute the thread

**Comment by a-velt** [comment]
Ok, thank you ! I found an alternative here : 
https://github.com/talgalili/heatmaply/issues/15

My code : 
In server.R : 
```
 output$heatmap_MANUAL <- renderPlotly({
 heatmaply(data_to_plot_heatmap_MANUAL_samples_log,colors = colorRampPalette(rev(brewer.pal(n = 7, name = "RdYlBu")))(100),labCol=colnames(data_to_plot_heatmap_MANUAL_samples_log))
      })
```

In ui.R:
`plotlyOutput("heatmap_MANUAL")`

My output : 
![image](https://user-images.githubusercontent.com/23554476/35983661-c4d8c68a-0cf2-11e8-803c-37455af2148b.png)

I think my labCol are too long, but I cannot reduce them, so maybe I have to play with margin etc. I will see.

When I tried to use shinyheatmaply, it generated me a shinyapp and not an output that I could integrate into my shiny app. I followed this tutorial : "shinyHeatmaply – a shiny app for creating interactive cluster heatmaps"

Thank you.
Best,
Amandine

**Comment by Alanocallaghan** [comment]
Hi Amandine, could you post a small example of how you generated the figure at the start? Without Shiny would be easier but with is fine.

**Comment by a-velt** [comment]
I don't know if I well understand, but I have a table with expression data (numerical value), with one line per gene and one sample per column, and I want to plos a heatmap of these expression in many genes and many samples, to compare them.

For that, I type the following command : 

`heatmaply(MY_TABLE,colors = colorRampPalette(rev(brewer.pal(n = 7, name = "RdYlBu")))(100),labCol=colnames(MY_TABLE))`

And I have a perfect heatmap, opened in a new html page : 
![image](https://user-images.githubusercontent.com/23554476/36090526-f9ea41fa-0fe0-11e8-8432-bdf4f72fd1a9.png)

But, when I want to insert this heatmap in my shiny app, I have a problem. I obtained that : 
![image](https://user-images.githubusercontent.com/23554476/36090580-300df6d2-0fe1-11e8-886c-f4c6e36af35d.png)

The heatmap doesn't properly scale my application. To put it in my application, I do that : 
`plotlyOutput("heatmap_CRIBI")`

But maybe there is an other trick to do that ?

**Comment by Alanocallaghan** [comment]
I believe that problem is largely related to the size of the window. Unfortunately plotly heatmaps don't scale particularly well when the window size is small. What might help is adjusting the font size of the column names (`fontsize_row`) and the margins (`margins`).

**Comment by Alanocallaghan** [comment]
Alternatively you could disable the tick labels and rely on the hover text using `showticklabels`

**Comment by a-velt** [comment]
I tested fontsize_row and margin, so it's better but I have a last question. 
It seems to me that the size of my heatmap never changes. Let me explain, the rectangle that represents my heatmap does'nt change, so if the number of variable on my Y axis increases, the labels are superimposed. How to change the size of the chart itself, not the margins?

Example of me new heatmap : 
![image](https://user-images.githubusercontent.com/23554476/36098912-0d4bf412-1001-11e8-9273-fcc0ad37b7fd.png)

If I increase the number of genes : 
![image](https://user-images.githubusercontent.com/23554476/36098951-257accd4-1001-11e8-88c9-4065de7b8e78.png)

**Comment by talgalili** [comment]
Did you consider using the abbreviate function but then supply the full
text to the hover?

**Comment by a-velt** [comment]
Yes, but the length of my xlab is no longer a problem. Or maybe it's because my labels are big that my chart is not growing? My problem is really that the rectangle that represents the heatmap doesn't change its size, that I have 50 genes or 1000 genes on the Y axis, its size is the same ... Is there a way to determine the size of the graph itself? I could change it dynamically depending on the number of genes selected on my application, by the user.

**Comment by Alanocallaghan** [comment]
The size of the graph is dependent on the HTML element it is generated within, which in turn is dependent on the size of the browser window it is in. heatmaply should do this automatically. You can test this out using [Chrome developer tools](https://developers.google.com/web/tools/chrome-devtools/device-mode/emulate-mobile-viewports)
